### PR TITLE
[Blazor] MSAL updates

### PR DESCRIPTION
**Pull request by javiercn**
* Updates MSAL to the latest version.
* Updates the typescript version used for compiling the code.
* Replaces history on navigation
* Adds logging for better diagnostics.
* Leverages crypto.randomUUID to generate a random string to use for the state parameter.
* Fixes support for requesting additional tokens interactively.
* Prevents triming MsalProviderOptions.

Addresses https://github.com/dotnet/aspnetcore/issues/43063, https://github.com/dotnet/aspnetcore/issues/39311, https://github.com/dotnet/aspnetcore/issues/29384

### CODE REVIEW

File: src/Components/WebAssembly/Authentication.Msal/src/Interop/AuthenticationService.ts [hunk 1]
```diff
@@ -128,72 +174,76 @@ class MsalAuthorizeService implements AuthorizeService {
...
         }
 
-        this._requestedScopes = scopes;
         const silentRequest = {
             redirectUri: this._settings.auth?.redirectUri,
             account: account,
             scopes: scopes || this._settings.defaultAccessTokenScopes
         };
 
+        this.debug(`Provisioning a token silently for scopes '${silentRequest.scopes}'`)
```
Comment (MackinnonBuck) [inline src/Components/WebAssembly/Authentication.Msal/src/Interop/AuthenticationService.ts:183]: Nit:
```suggestion
        this.debug(`Provisioning a token silently for scopes '${silentRequest.scopes}'`);
```

File: src/Components/WebAssembly/Authentication.Msal/src/Interop/AuthenticationService.ts [hunk 2]
```diff
@@ -321,16 +369,22 @@ class MsalAuthorizeService implements AuthorizeService {
...
+            return res;
+        } catch (error) {
             if (this.isMsalError(error)) {
                 return this.error(error.errorMessage);
             } else {
-                return this.error(error);
+                return this.error((error as Error).message);
             }
-        })
+        };
```
Comment (MackinnonBuck) [inline src/Components/WebAssembly/Authentication.Msal/src/Interop/AuthenticationService.ts:387]: Silly nit:
```suggestion
        }
```

File: src/Components/WebAssembly/Authentication.Msal/src/Interop/AuthenticationService.ts [hunk 3]
```diff
@@ -43,6 +43,54 @@ interface AuthorizeService {
...
+export interface AuthenticationContext {
+    state?: unknown;
+    interactiveRequest: InteractiveAuthenticationRequest;
+}
+
+export interface InteractiveAuthenticationRequest {
+    interaction: string;
+    scopes?: string[];
+    additionalRequestParameters?: { [key: string]: any };
+};
```
Comment (MackinnonBuck) [inline src/Components/WebAssembly/Authentication.Msal/src/Interop/AuthenticationService.ts:92]:
```suggestion
}
```

File: src/Components/WebAssembly/Authentication.Msal/src/Interop/AuthenticationService.ts [hunk 4]
```diff
@@ -51,14 +99,35 @@ interface AuthorizeServiceConfiguration extends Msal.Configuration {
...
+                location.replace(url);
+                return false;
+            },
+            async navigateExternal(url: string, options: Msal.NavigationOptions): Promise<boolean> {
+                // We always replace the URL
+                _logger.log(LogLevel.Trace, `Navigating to ${url}`);
+                location.replace(url);
+                return false;
+            }
+        }
```
Comment (MackinnonBuck) [inline src/Components/WebAssembly/Authentication.Msal/src/Interop/AuthenticationService.ts:125]:
```suggestion
        };
```

File: src/Components/WebAssembly/Authentication.Msal/src/Interop/AuthenticationService.ts [hunk 5]
```diff
@@ -43,6 +43,54 @@ interface AuthorizeService {
...
+    Trace = 0,
+    Debug = 1
+}
+
+interface JavaScriptLoggingOptions {
+    debugEnabled: boolean;
+    traceEnabled: boolean;
+}
+
+export class Logger {
```
Comment (MackinnonBuck) [inline src/Components/WebAssembly/Authentication.Msal/src/Interop/AuthenticationService.ts:58]: Nit: I'm probably missing something, but why are these types `export`ed? Are they going to be consumed elsewhere?

**Review by MackinnonBuck — APPROVED** [review]
LGTM.

Nothing but nits.

File: src/Components/WebAssembly/Authentication.Msal/src/Interop/AuthenticationService.ts (hunk 5 above)
Comment (javiercn) [reply src/Components/WebAssembly/Authentication.Msal/src/Interop/AuthenticationService.ts:58]: They are not consumed outside the lib, they could technically not be exported, but it does not matter.

**Comment by javiercn** [comment]
@Pilchie this should be good to go.

**Comment by Pilchie** [comment]
Approved for .NET 7 RC2.

**Pull request merged**